syntax = "proto3";

package universe;

service Controller {
    rpc JoinGroup (JoinGroupRequest) returns (JoinGroupReply) {}
    rpc LeaveGroup (LeaveGroupRequest) returns (LeaveGroupReply) {}
    rpc CacheWorkerMap (CacheWorkerMapRequest) returns (CacheWorkerMapReply) {}

    rpc ReadFault(FaultRequest) returns (FaultReply) {}
    rpc WriteFault(FaultRequest) returns (FaultReply) {}
    rpc EndFault(FaultRequest) returns (FaultReply) {}
}

message JoinGroupRequest {
    int32  rank = 1;
    repeated uint32 qpn = 2;
    uint32 lid = 3;  // only lower 16 bits used
    uint64 addr = 4;
    uint32 key = 5;
    bytes  gid = 6;  // uint8[16], currently unused
    uint64 size = 7;
}

message JoinGroupReply {
    int32 err_code = 1;  // reserved
    int32 num_of_procs = 2;
}

message CacheWorkerMapRequest {
    repeated int32 want_rank = 1;
}

message CacheWorkerMapReply {
    int32 err_code = 1;  // reserved
    repeated JoinGroupRequest worker = 2;
}

message LeaveGroupRequest {
    int32 rank = 1;
}

message LeaveGroupReply {
    int32 err_code = 1;  // reserved
}

message FaultRequest {
    int32 rank = 1;
    uint64 addr = 2;
}

message FaultReply {
    int32 err_code = 1;  // reserved
    repeated int32 cached_rank = 2;
}
